labgrid/driver/sshdriver.py

```diff
@@ -82,10 +83,8 @@ def _start_own_master_once(self, timeout):
         control = os.path.join(
             self.tmpdir, 'control-{}'.format(self.networkservice.address)
         )
-        # use sshpass if we have a password
-        args = ["sshpass", "-e"] if self.networkservice.password else []
 
-        args += ["ssh", "-f", *self.ssh_prefix, "-x", "-o", "ConnectTimeout={}".format(timeout),
+        args = ["ssh", "-f", *self.ssh_prefix, "-x", "-o", "ConnectTimeout={}".format(timeout),
                  "-o", "ControlPersist=300", "-o",
                  "UserKnownHostsFile=/dev/null", "-o", "StrictHostKeyChecking=no",
                  "-o", "ServerAliveInterval=15", "-MN", "-S", control.replace('%', '%%'), "-p",
```

Comment:
Can we remove the ServerAliveInterval argument and leave it for the ~/.ssh/config? Because of the missing
ServerAliveCountMax value our session logges out after 60s.

Reply:
This line is missing a `["-F", "none"]`to ignore the user configuration file. ServerAliveInterval is correct here, since the keepalive started in `on_activate` should keep the socket open as long as your device is reachable. This breaks if you restart the target unexpectedly without deactivating the SSHDriver first and is a usage error.

Reply:
We explicitly add `-F none` to `self.ssh_prefix` after starting the master. Are you sure that was not intentional and a change wouldn't break user configs (such as per device keys)?

Reply:
You mean the usecase where you have target keys configured in you SSH config? AFAICS this is a mistake and keys should be directly configured in the YAML. I am pretty sure this is a leftover when we removed the support for user supplied masters altogether and this didn't work for several years. I don't think we should support per user keys configured in the SSH config and instead want them to be defined explicitly in the YAML file.

Reply:
OK. But then we should move the addition of `-F none` up instead of adding it to individual calls.

Reply:
Ok, I found the _start_keepalive function, but I had to add the address to the args in order for the "cat" command to be executed correctly. Unfortunately, I am still being logged out after 60s.

    def _start_keepalive(self):
            """Starts a keepalive connection via the own or external master."""
             args = ["ssh", *self.ssh_prefix, self.networkservice.address, "cat"]
`

Reply:
> This line is missing a `["-F", "none"]`to ignore the user configuration file. ServerAliveInterval is correct here, since the keepalive started in `on_activate` should keep the socket open as long as your device is reachable. This breaks if you restart the target unexpectedly without deactivating the SSHDriver first and is a usage error.

@Emantor the start_keepalive keeps the socket open, but not the labgrid-client connection, right? I think the ServerAliveInterval could be removed, or increased, to get a higher connection timeout for the client.

Reply:
@Emantor ok, never mind. We are using a dropbear server, which doesn't seem to work very well together with openssh ServerAliveInterval.

Reply:
Is that a known issue for dropbear? I'd prefer to keep the ServerAlive, as otherwise we'd need to implement "ping" ourselves via the cat...

Reply:
It looks like the openssh ServerAliveInterval is just not supported by dropbear. I found a dropbear server option that allows to increase the idle timeout, but was not able to keep the connection open permanently. I am fine with how it works right now, as long as the connections stays open for longer than a minute.

Reply:
We could add a driver option to support dropbear. From reading the dropbear changelog, it simply does not respond to the server keepalive messages, thus the client thinks the server is dead and terminates the connection, AFAICS. So for dropbear we would remove the `ServerAliveInterval` and maybe enable the TCP keepalive only.

Reply:
Sounds like a good idea :smile: I can add that in another patch.

Reply:
> We could add a driver option to support dropbear. From reading the dropbear changelog, it simply does not respond to the server keepalive messages, thus the client thinks the server is dead and terminates the connection, AFAICS. So for dropbear we would remove the `ServerAliveInterval` and maybe enable the TCP keepalive only.

@Emantor I tested by adding a server_type attribute to distinguish between openssh and dropbear and I played with removing the ServerAliveInterval and adding TCPKeepAlive (yes and no). I even removed all occurrences of ControlPersist=300 and also tried by setting yes and no. However without the ServerAliveIntervalI I am being disconnected after exactly 5 minutes. This is regardless of using a proxy or not. I guess the openssh client, dropbear server combo needs some more testing.

The client disconnection log looks like this. The server says the the client exited correctly.

```
...
debug2: channel 4: rcvd eof
debug2: channel 4: output open -> drain
debug2: channel 4: obuf empty
debug2: channel 4: chan_shutdown_write (i0 o1 sock -1 wfd 12 efd 13 [write])
debug2: channel 4: output drain -> closed
debug3: receive packet: type 98
debug1: client_input_channel_req: channel 4 rtype exit-status reply 0
debug3: mux_exit_message: channel 4: exit message, exitval 0
debug3: receive packet: type 97
debug2: channel 4: rcvd close
debug2: channel 4: chan_shutdown_read (i0 o3 sock -1 wfd 11 efd 13 [write])
debug2: channel 4: input open -> closed
debug3: channel 4: will not send data after close
debug2: channel 4: send close
debug3: send packet: type 97
debug2: channel 4: is dead
debug2: channel 4: gc: notify user
debug3: mux_master_session_cleanup_cb: entering for channel 4
debug2: channel 3: rcvd close
debug2: channel 3: output open -> drain
debug2: channel 3: chan_shutdown_read (i0 o1 sock 10 wfd 10 efd -1 [closed])
debug2: channel 3: input open -> closed
debug2: channel 4: gc: user detached
debug2: channel 4: is dead
debug2: channel 4: garbage collecting
debug1: channel 4: free: client-session, nchannels 5
debug3: channel 4: status: The following connections are open:
  #1 mux-control (t16 r2 i0/0 o0/0 e[closed]/0 fd 6/6/-1 sock 6 cc -1)
  #2 client-session (t4 r0 i0/0 o0/0 e[write]/0 fd 7/8/9 sock -1 cc 1)
  #3 mux-control (t16 nr0 i3/0 o1/16 e[closed]/0 fd 10/10/-1 sock 10 cc -1)
  #4 client-session (t4 r1 i3/0 o3/0 e[write]/0 fd -1/-1/13 sock -1 cc -1)

debug2: channel 3: obuf empty
debug2: channel 3: chan_shutdown_write (i3 o1 sock 10 wfd 10 efd -1 [closed])
debug2: channel 3: output drain -> closed
debug2: channel 3: is dead (local)
debug2: channel 3: gc: notify user
debug3: mux_master_control_cleanup_cb: entering for channel 3
debug2: channel 3: gc: user detached
debug2: channel 3: is dead (local)
debug2: channel 3: garbage collecting
debug1: channel 3: free: mux-control, nchannels 4
debug3: channel 3: status: The following connections are open:
  #1 mux-control (t16 r2 i0/0 o0/0 e[closed]/0 fd 6/6/-1 sock 6 cc -1)
  #2 client-session (t4 r0 i0/0 o0/0 e[write]/0 fd 7/8/9 sock -1 cc 1)
  #3 mux-control (t16 nr0 i3/0 o3/0 e[closed]/0 fd 10/10/-1 sock 10 cc -1)

debug3: send packet: type 1
client_loop: send disconnect: Broken pipe
```